// @progress/kendo-react-grid/dist/es/drag/ColumnDraggable.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import * as React from 'react';
import Draggable from '@telerik/kendo-draggable';
/**
 * @hidden
 */
var ColumnDraggable = /** @class */ (function (_super) {
    __extends(ColumnDraggable, _super);
    function ColumnDraggable() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.element = null;
        _this.draggable = new Draggable({
            press: function (event) {
                if (_this.element && _this.props.pressHandler) {
                    _this.props.pressHandler(event, _this.element);
                }
            },
            drag: function (event) {
                if (_this.element && _this.props.dragHandler) {
                    _this.props.dragHandler(event, _this.element);
                }
            },
            release: function (event) {
                if (_this.element && _this.props.releaseHandler) {
                    _this.props.releaseHandler(event);
                }
            }
        });
        return _this;
    }
    ColumnDraggable.prototype.componentDidMount = function () {
        if (this.element) {
            this.draggable.bindTo(this.element);
        }
    };
    ColumnDraggable.prototype.componentWillUnmount = function () {
        this.draggable.destroy();
    };
    ColumnDraggable.prototype.render = function () {
        var _this = this;
        return (React.createElement("tr", { ref: function (tr) { return _this.element = tr; } }, this.props.children));
    };
    return ColumnDraggable;
}(React.PureComponent));
export { ColumnDraggable };
