// @progress/kendo-react-grid/dist/es/drag/ColumnResizer.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import * as React from 'react';
import Draggable from '@telerik/kendo-draggable';
/**
 * @hidden
 */
var ColumnResizer = /** @class */ (function (_super) {
    __extends(ColumnResizer, _super);
    function ColumnResizer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.element = null;
        _this.draggable = new Draggable({
            drag: function (event) {
                if (_this.element) {
                    _this.props.resize(event, _this.element, false);
                }
            },
            release: function (event) {
                if (_this.element) {
                    _this.props.resize(event, _this.element, true);
                }
            }
        });
        return _this;
    }
    ColumnResizer.prototype.componentDidMount = function () {
        if (this.element) {
            this.draggable.bindTo(this.element);
        }
    };
    ColumnResizer.prototype.componentWillUnmount = function () {
        this.draggable.destroy();
    };
    ColumnResizer.prototype.render = function () {
        var _this = this;
        var resizeStyle = {
            cursor: 'col-resize',
            display: 'block',
            height: '1000%',
            position: 'absolute',
            right: 0,
            top: 0,
            width: '.5em'
        };
        return (React.createElement("span", { className: "k-column-resizer", draggable: false, style: resizeStyle, ref: function (el) { _this.element = el; } }));
    };
    return ColumnResizer;
}(React.PureComponent));
export { ColumnResizer };
